eth/stagedsync/ots_indexer_log.go

```diff
@@ -0,0 +1,344 @@
+// can contain which addresses this log entry touches in a token transfer indexer.
+//
+// An instance of this interface is meant to be reused, so it can contain caches to speed up further
+// analysis.
+type LogAnalyzer[T any] interface {
+	// Given a log entry (there may be others in the same tx, here we analyze 1 specific log entry),
+	// does it match the criteria the implementation is suposed to analyze?
+	//
+	// Return nil means it doesn't pass the criteria and it shouldn't be indexed.
+	Inspect(tx kv.Tx, l *types.Log) (*T, error)
+}
+
+// Handles log indexer lifecycle.
+type LogIndexerHandler[T any] interface {
+	ResourceAwareIndexHandler
+
+	// Given a tx that must be indexed, handles all logs that caused the matching.
+	HandleMatch(match *TxMatchedLogs[T])
+}
+
+func runConcurrentLogIndexerExecutor[T any](db kv.RoDB, tx kv.RwTx, blockReader services.FullBlockReader, startBlock, endBlock uint64, isShortInterval bool, logEvery *time.Ticker, ctx context.Context, s *StageState, analyzer LogAnalyzer[T], handler LogIndexerHandler[T]) (uint64, error) {
+	if !isShortInterval {
+		log.Info(fmt.Sprintf("[%s] Using concurrent executor", s.LogPrefix()))
+	}
+
+	// Scans logs bucket
+	logs, err := tx.Cursor(kv.Log)
+	if err != nil {
+		return startBlock, err
+	}
+	defer logs.Close()
+
+	// Tracks how many txs finished analysis so far
+	totalTx := atomic.NewUint64(0)
+
+	// Tracks how many txs finished analysis with a match so far
+	totalMatch := atomic.NewUint64(0)
+
+	// Each worker processes all logs from 1 ethTx each time
+	workers := estimate.AlmostAllCPUs()
+	proberCh := make(chan *TxLogs[T], workers*3)
+	matchCh := make(chan *TxMatchedLogs[T], workers*3+workers) // must be >= inCh buffer size + n. of workers, otherwise can deadlock
+
+	g, gCtx := errgroup.WithContext(ctx)
+	for i := 0; i < workers; i++ {
+		createLogAnalyzerWorker(g, gCtx, db, analyzer, proberCh, matchCh, totalMatch, totalTx)
+	}
+
+	// Process control
+	flushEvery := time.NewTicker(bitmapsFlushEvery)
+	defer flushEvery.Stop()
+
+	// Get all blocks in [startBlock, endBlock] that contain at least 1 TRANSFER_TOPIC occurrence
+	blocks, err := newBlockBitmapFromTopic(tx, startBlock, endBlock, TRANSFER_TOPIC)
+	if err != nil {
+		return startBlock, err
+	}
+	defer bitmapdb.ReturnToPool(blocks)
+
+L:
+	// Iterate each block that contains a Transfer() event
+	for it := blocks.Iterator(); it.HasNext(); {
+		blockNum := uint64(it.Next())
+
+		// Avoid recalculating txid from the block basetxid for each match
+		baseTxId, err := blockReader.BaseTxIdForBlock(ctx, tx, blockNum)
+		if err != nil {
+			return startBlock, err
+		}
+
+		// Inspect each tx's logs
+		logPrefix := hexutility.EncodeTs(blockNum)
+		for k, v, err := logs.Seek(logPrefix); k != nil && bytes.HasPrefix(k, logPrefix); k, v, err = logs.Next() {
+			if err != nil {
+				return startBlock, err
+			}
+
+			txLogs := newTxLogsFromRaw[T](blockNum, baseTxId, k, v)
```

Comment:
`baseTxId` - it's internal technical field: auto_increment id. also it's ok to have different id on different nodes. probably don't need expose it for user.

Reply:
It looks like when ots2-alpha6-wip was force-pushed from 1fdbc12 to 57a48ef last week, this PR's diff suddenly included all the changes from 1fdbc12.

So @wmitsuda - I think this comment is for you.

I'll rebase this PR on top of 57a48ef.